### PR TITLE
[sinks] Remove the old reuse-topic option

### PR DESCRIPTION
**Pull request by bkirwi**
We no longer plan to support the old behaviour of creating a new topic every time that materialize restarts, so we no longer need a configuration option to opt in to the newer behaviour.

This also allows us to clean out a number of parameters / degrees of freedom that are no longer used. I've tried to stick to deleting stuff that we're pretty sure we'll never use again, like the topic suffixing machinery.

### Motivation

Known-desirable feature #14372!

### Tips for reviewer

The first commit cleans out the code, and the second updates the tests. Most of the changes are pretty mechanical, but I've noted one or two that might be more surprising.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Removes the `reuse_topic` option for sinks; the behaviour it opted into is now the default, and the old behaviour is unsupported.

### CODE REVIEW

File: src/adapter/src/sink_connection.rs
Comment (brennanvincent) [inline src/adapter/src/sink_connection.rs:297]: Can this parameter be removed?

File: src/adapter/src/sink_connection.rs
Comment (brennanvincent) [inline src/adapter/src/sink_connection.rs:211]: Why do we need to clone here?

File: src/adapter/src/sink_connection.rs
Comment (brennanvincent) [inline src/adapter/src/sink_connection.rs:231]: Idem

**Review by brennanvincent — COMMENTED** [review]
LGTM in principle but you're better off getting someone from the Storage team for final review!

**Comment by bkirwi** [comment]
@brennanvincent - Hey, looks like you've caught me mid-draft here... will definitely sort out those issues by the time this is ready to go!

**Comment by brennanvincent** [comment]
No worries @bkirwi , somehow I saw this and missed that it was a draft!

File: src/adapter/src/sink_connection.rs
Comment (bkirwi) [reply src/adapter/src/sink_connection.rs:294]: We _could_, but IIUC there's some chance we'll re-introduce a not-exactly-once option in the future, so we're not quite as keen to rip that stuff out. (Whereas we're pretty confident that we won't ever do the topic-nonce-suffix thing again.)

File: test/upgrade/create-in-current_source-kafka-sink.td
Comment (bkirwi) [inline test/upgrade/create-in-current_source-kafka-sink.td:18]: These messages are checked in the corresponding `check` test.

Normally `verify` consumes the messages from the topic, so it's not totally obvious to me why this used to work...

**Comment by bkirwi** [comment]
The build is currently broken because `main` is... I'll merge it in once that's fixed.

File: src/sql/src/plan/statement/ddl.rs
Comment (sploiselle) [inline src/sql/src/plan/statement/ddl.rs:1905]: If it only supports true, we should just remove the option entirely.

File: src/storage/src/types/sinks.rs
Comment (sploiselle) [inline src/storage/src/types/sinks.rs:423]: asking out of curiosity: why don't we just remove/deprecate this from the proto definition?

Comment (bkirwi) [reply src/storage/src/types/sinks.rs:423]: Same answer for both, basically: a halfhearted attempt to give a hard error if folks are still using the old feature, instead of silently doing the wrong thing.

But I don't think we really care, and we're certainly breaking a lot this week anyways! So I'll clean them both out.

**Review by sploiselle — APPROVED** [review]

**Comment by bkirwi** [comment]
🙇‍♂️

**Comment by aljoscha** [comment]
Just saw the branch name ... 😂

**Comment by cjubb39** [comment]
@bkirwi very exciting that this is being removed!!

Should we go ahead and remove references to the `reuse_topic` option from the docs?  You'd find those in `doc/user/content/sql/create-sink.md`

**Comment by bkirwi** [comment]
Yeah, we definitely need to do a pass through the docs... a lot will have changed, down to the syntax diagram. I'm not actually sure when's the right time to do that / what the process is, but I've been waiting until the dust settles a bit... if you think now's the time we can get started though!

**Comment by cjubb39** [comment]
Fair enough -- I think booting the `reuse_topic` option shouldn't take too long since we know it for sure it's gone forever -- but we could defer if you like and do it all at once

**Pull request merged**
